gossip/src/cluster_info.rs

```diff
@@ -3215,6 +3215,97 @@ mod tests {
         ));
     }
 
+    fn rand_port_range() -> (u16, u16) {
+        let width = 100;
+        let start = thread_rng().gen_range(1024..(65535 - width));
+        (start, start + width)
+    }
+
+    #[test]
+    fn test_duplicate_ports_single_bind() {
+        solana_logger::setup();
+
+        let ip = Ipv4Addr::LOCALHOST;
+        let bind_ip_addr = IpAddr::V4(ip);
+        let node = Node::new_single_bind(
+            &Pubkey::new_unique(),
+            &socketaddr!(ip, 0),
+            rand_port_range(),
+            bind_ip_addr,
+        );
+
+        check_duplicate_ports(&node.info, true);
+    }
+
+    #[test]
+    fn test_duplicate_ports_external_ip() {
+        solana_logger::setup();
+
+        let ip = Ipv4Addr::LOCALHOST;
+        let config = NodeConfig {
+            gossip_addr: socketaddr!(ip, 0),
+            port_range: rand_port_range(),
+            bind_ip_addr: IpAddr::V4(ip),
+            public_tpu_addr: None,
+            public_tpu_forwards_addr: None,
+        };
+
+        let node = Node::new_with_external_ip(&Pubkey::new_unique(), config);
+
+        check_duplicate_ports(&node.info, false);
+    }
+
+    #[test]
+    fn test_duplicate_ports_local() {
+        solana_logger::setup();
+        let node = Node::new_localhost();
+
+        check_duplicate_ports(&node.info, true);
+    }
+
+    fn check_duplicate_ports(info: &ContactInfo, rpc: bool) {
+        let mut ports = HashSet::new();
+        let s = info.gossip().unwrap();
+        assert!(ports.insert(s.port()));
```

Comment:
@behzadnouri maybe there is a cleaner way to do this?

Reply:
If we only need to check my node's contact-info (as opposed to another node's contact-info), we can enforce that `socket` address is not duplicate here:
https://github.com/anza-xyz/agave/blob/191a997c5/gossip/src/contact_info.rs#L303
when setting socket addresses on the contact-info.

Reply:
something like this:
```diff
diff --git a/gossip/src/contact_info.rs b/gossip/src/contact_info.rs
index b3ca9c94a7..6533d98336 100644
--- a/gossip/src/contact_info.rs
+++ b/gossip/src/contact_info.rs
@@ -44,6 +44,8 @@ const SOCKET_CACHE_SIZE: usize = SOCKET_TAG_TVU_QUIC as usize + 1usize;
 pub enum Error {
     #[error("Duplicate IP address: {0}")]
     DuplicateIpAddr(IpAddr),
+    #[error("Duplicate port: {0}")]
+    DuplicatePort(/*port:*/ u16),
     #[error("Duplicate socket: {0}")]
     DuplicateSocket(/*key:*/ u8),
     #[error("Invalid IP address index: {index}, num addrs: {num_addrs}")]
@@ -301,6 +303,22 @@ impl ContactInfo {
     }
 
     pub fn set_socket(&mut self, key: u8, socket: SocketAddr) -> Result<(), Error> {
+        // Verify that the socket is not duplicate.
+        let _ = self.sockets.iter().try_fold(0u16, |port, entry| {
+            let Some(port) = port.checked_add(entry.offset) else {
+                return Err(Error::PortOffsetsOverflow);
+            };
+            let Some(&addr) = self.addrs.get(usize::from(entry.index)) else {
+                return Err(Error::InvalidIpAddrIndex {
+                    index: entry.index,
+                    num_addrs: self.addrs.len(),
+                });
+            };
+            if addr == socket.ip() && port == socket.port() && entry.key != key {
+                return Err(Error::DuplicatePort(port));
+            }
+            Ok(port)
+        })?;
         sanitize_socket(&socket)?;
         // Remove the old entry associated with this key (if any).
         self.remove_socket(key);
```